lib/CL/devices/common_driver.c

```diff
@@ -449,13 +453,97 @@ pocl_driver_svm_fill (cl_device_id dev, void *__restrict__ svm_ptr,
  */
 
 #ifdef ENABLE_LLVM
+
+/* if some SPIR-V spec constants were changed, use llvm-spirv --spec-const=...
+ * to generate new LLVM bitcode from SPIR-V */
+static int
+pocl_regen_spirv_binary (cl_program program, cl_uint device_i)
+{
+#ifdef LLVM_SPIRV
+  int errcode = CL_SUCCESS;
+
+  int spec_constants_changed = 0;
+
+  for (unsigned i = 0; i < program->num_spec_consts; ++i)
+    spec_constants_changed += program->spec_const_is_set[i];
```

Comment:
Does this regenerate when the constant is set but to a different value? Should we just add the specialization constants to the hash?

Reply:
Yes, the current value (last set) is used at every clBuildProgram. If the user sets the constant and doesn't call clBuildProgram then it will use the old value, but that's the user's fault. Not sure about adding to the build hash. It should work without as is, but in a sense these are like build options, and IIRC we add the build options to the build hash of non-spirv programs, so it probably should be included.

Reply:
That way the specialized SPIR-Vs and associated generated codes would get cached correctly?